# Formato de cadena

El formato de cadena le permite agregar elementos en una cadena en lugar de intentar encadenar elementos mediante comas o concatenación de cadenas. Como comparación rápida, considere:

     jugador = 'Thomas'
     puntos = 33
    
     'Anoche,' + jugador + 'anotó' + str (puntos) + 'puntos'. # concatenación
    
     f'La noche pasada, {player} anotó {points} puntos '. # formato de cadena


Hay tres formas de realizar el formateo de cadenas.
* El método más antiguo implica marcadores de posición que utilizan el carácter de módulo **`%`**.
* Una técnica mejorada utiliza el método de cadena **`.format ()`**.
* El método más nuevo, introducido con Python 3.6, utiliza cadenas literales formateadas, llamadas **f-strings**.

Dado que probablemente encontrará las tres versiones en el código de otra persona, describimos cada una de ellas aquí.

## Formateo con marcadores de posición
Puede usar <code>% s </code> para agregar cadenas en sus declaraciones de impresión. El módulo `%` se conoce como un "operador de formato de cadena".

In [1]:
print("Voy agregar %s aqui." %'algo')

Voy agregar algo aqui.


Puede pasar varios elementos colocándolos dentro de una tupla después del operador `%`.

In [2]:
print("Voy a insertar %s textos aquí y %s texto aquí." % ('algunos', 'más'))

Voy a insertar algunos textos aquí y más texto aquí.


También puede pasar nombres de variables:

In [4]:
x, y = 'algunos', 'más'
print ("Voy a insertar %s textos aquí y %s texto aquí." % (x, y))

Voy a insertar algunos textos aquí y más texto aquí.


### Métodos de conversión de formato.
Cabe señalar que dos métodos <code>%s </code> y <code>%r </code> convierten cualquier objeto de Python en una cadena usando dos métodos separados: `str()` y `repr()`. Aprenderemos más sobre estas funciones más adelante en el curso, pero debe tener en cuenta que `%r` y` repr() `entregan la * representación de cadena * del objeto, incluidas las comillas y los caracteres de escape.

In [13]:
print ('Dijo que se llamaba %s.' %'Fred')
print ('Dijo que se llamaba %r.' %'Fred')

Dijo que se llamaba Fred.
Dijo que se llamaba 'Fred'.


Como otro ejemplo, `\ t` inserta una pestaña en una cadena.

In [15]:
print ('Una vez pesqué un pez %s.' %'este \tgrande')
print ('Una vez pesqué un pez %r.' %'este \tgrande')

Una vez pesqué un pez este 	grande.
Una vez pesqué un pez 'este \tgrande'.


El operador `%s` convierte todo lo que ve en una cadena, incluidos enteros y flotantes. El operador `%d` convierte números a enteros primero, sin redondear. Tenga en cuenta la diferencia a continuación:

In [5]:
print ('Escribí %s programas hoy.' %3.75)
print ('Escribí %d programas hoy.' %3.75)

Escribí 3.75 programas hoy.
Escribí 3 programas hoy.


### Relleno y precisión de números de punto flotante
Los números de coma flotante usan el formato <code>%5.2f </code>. Aquí, <code>5</code> sería el número mínimo de caracteres que debe contener la cadena; estos se pueden rellenar con espacios en blanco si el número completo no tiene tantos dígitos. Junto a esto, <code>.2f</code> representa cuántos números mostrar después del punto decimal. Veamos algunos ejemplos:

In [7]:
print ('Números de coma flotante:%5.2f'% (13.144))

Números de coma flotante:13.14


In [8]:
print ('Números de coma flotante:%1.0f'% (13.144))

Números de coma flotante:13


In [9]:
print ('Números de coma flotante:%1.5f'% (13.144))

Números de coma flotante:13.14400


In [11]:
print ('Números de coma flotante:%10.2f'% (13.144))

Números de coma flotante:     13.14


In [12]:
print ('Números de coma flotante:% 25.2f'% (13.144))

Números de coma flotante:                    13.14


Para obtener más información sobre el formato de cadenas con marcadores de posición, visite https://docs.python.org/3/library/stdtypes.html#old-string-formatting

### Formateo múltiple
Nada prohíbe el uso de más de una herramienta de conversión en la misma declaración de impresión:

In [13]:
print('Primero: %s, Segundo: %5.2f, Tercero: %r' %('Hola!',3.1415,'!adiós!'))

Primero: Hola!, Segundo:  3.14, Tercero: '!adiós!'


## Formateo con el método `.format ()`
Una mejor manera de formatear objetos en sus cadenas para declaraciones de impresión es con el método de cadena `.format ()`. La sintaxis es:

     'Cadena aquí {} y luego también {}'. Format ('algo1', 'algo2')
    
Por ejemplo:

In [14]:
print('Esto es una cadena con una {}'.format('inserción'))

Esto es una cadena con una inserción


### El método .format () tiene varias ventajas sobre el método de marcador de posición %s:

#### 1. Los objetos insertados se pueden llamar por posición de índice:

In [28]:
print ('El {2} {1} {0}'.format('rápido', 'marrón', 'zorro'))

El zorro marrón rápido


#### 2. A los objetos insertados se les pueden asignar palabras clave:

In [15]:
print ('Primer objeto: {a}, Segundo objeto: {b}, Tercer objeto: {c}'.format (a = 1, b = 'Dos', c = 12.3))

Primer objeto: 1, Segundo objeto: Dos, Tercer objeto: 12.3


#### 3. Los objetos insertados se pueden reutilizar, evitando la duplicación:

In [16]:
print ('Un %s guardado es un %s ganado.'% ('centavo', 'centavo'))
# vs.
print ('Un {p} guardado es un {p} ganado.'. format (p = 'centavo'))

Un centavo guardado es un centavo ganado.
Un centavo guardado es un centavo ganado.


### Alineación, relleno y precisión con `.format ()`
Dentro de las llaves puede asignar longitudes de campo, alineaciones izquierda / derecha, parámetros de redondeo y más

In [36]:
print ('{0:8} | {1:9}'.format('Fruta', 'Cantidad'))
print ('{0:8} | {1:9}'.format('Manzanas', 3.))
print ('{0:8} | {1:9}'.format('Naranjas', 10))

Fruta    | Cantidad 
Manzanas |       3.0
Naranjas |        10


Por defecto, `.format ()` alinea el texto a la izquierda y los números a la derecha. Puede pasar un `<`, `^` o `>` opcional para establecer una alineación a la izquierda, al centro o a la derecha:

In [39]:
print ('{0:<8} | {1:^8} | {2:>8}'.format('Izquierda', 'Centro', 'Derecha'))
print ('{0:<8} | {1:^8} | {2:>8}'.format(11,22,33))

Izquierda |  Centro  |  Derecha
11       |    22    |       33


Puede preceder al operador de asignación con un caracter de relleno

In [17]:
print ('{0:=<8} | {1:-^8} | {2:.>8}'.format('Izquierda', 'Centro', 'Derecha'))
print ('{0:=<8} | {1:-^8} | {2:.>8}'.format(11,22,33))

Izquierda | -Centro- | .Derecha
11====== | ---22--- | ......33


Los anchos de campo y la precisión de flotación se manejan de manera similar a los marcadores de posición. Las siguientes dos declaraciones impresas son equivalentes:

In [18]:
print ('Este es mi caracter número de diez, numero y dos-decimales:%10.2f' %13.579)
print ('Este es mi caracter número de diez, numero y dos-decimales:{0:10.2f}'.format(13.579))

Este es mi caracter número de diez, numero y dos-decimales:     13.58
Este es mi caracter número de diez, numero y dos-decimales:     13.58


Tenga en cuenta que hay 5 espacios después de los dos puntos y 5 caracteres ocupados por 13.58, para un total de diez caracteres.

Para obtener más información sobre el método string `.format ()`, visite https://docs.python.org/3/library/string.html#formatstrings

## Literales de cadena formateados (f-strings)

Introducido en Python 3.6, f-strings ofrecen varios beneficios sobre el antiguo método de cadena **`.format ()`** descrito anteriormente. Por un lado, puede traer variables externas inmediatamente a la cadena en lugar de pasarlas como argumentos a través de `.format (var)`.

In [19]:
nombre = 'Fred'

print (f"Dijo que su nombre es {nombre}.")

Dijo que su nombre es Fred.


Pase `!r` para obtener la representación de cadena:

In [20]:
print (f"Dijo que su nombre es {nombre!r}")

Dijo que su nombre es 'Fred'


#### El formato flotante sigue `" resultado: {valor: {ancho}. {Precisión}} "`

Donde con el método `.format ()` puede ver `{valor: 10.4f}`, con f-strings esto puede convertirse en `{valor: {10}. {6}}`


In [21]:
num = 23.45678
print ("Mi número de diez caracteres, cuatro decimales es: {0: 10.4f}".format(num))
print (f"Mi número de 10 caracteres, cuatro números decimales es: {num: {10}.{6}}")

Mi número de diez caracteres, cuatro decimales es:    23.4568
Mi número de 10 caracteres, cuatro números decimales es:    23.4568


Tenga en cuenta que con f-strings, * precisión * se refiere al número total de dígitos, no solo a los que siguen al decimal. Esto encaja más de cerca con la notación científica y el análisis estadístico. Desafortunadamente, las cadenas f no se rellenan a la derecha del decimal, incluso si la precisión lo permite:

In [22]:
num = 23.45
print ("Mi número de diez caracteres, cuatro decimales es: {0:10.4f}".format(num))
print (f"Mi número de 10 caracteres, cuatro números decimales es: {num: {10}.{6}}")

Mi número de diez caracteres, cuatro decimales es:    23.4500
Mi número de 10 caracteres, cuatro números decimales es:      23.45


Si esto se vuelve importante, siempre puede usar la sintaxis del método `.format ()` dentro de una cadena f:

In [23]:
num = 23.45
print ("Mi número de diez caracteres, cuatro decimales es: {0:10.4f}".format(num))
print (f"Mi 10 carácter, cuatro números decimales es: {num: 10.4f}")

Mi número de diez caracteres, cuatro decimales es:    23.4500
Mi 10 carácter, cuatro números decimales es:    23.4500


Para obtener más información sobre los literales de cadena formateados, visite https://docs.python.org/3/reference/lexical_analysis.html#f-strings

¡Eso es lo básico del formateo de cadenas!